[rocprofiler-systems] [Refactor] Use PMC collectors instead of the previous amd-smi approach

adjordje-amd opened the pull request:
## Motivation

The current AMD SMI implementation faces several challenges related to its maintainability and architecture, which this refactor aims to address:

Monolithic Design: The original amd_smi.cpp file, comprising 1354 lines, merges sampling, caching, and output logic. This tight coupling makes it challenging to modify or extend individual components.

Lack of Testability: Due to the integration with AMD SMI hardware APIs, the implementation is difficult to unit test and necessitates real hardware for verification.

Limited Flexibility: Introducing support for new output formats, such as RocPD alongside Perfetto, involves substantial code duplication.

Accumulated Bugs: The implementation contains several persistent bugs

The refactoring process aims to tackle these issues by replacing the monolithic setup with a contemporary, policy-based design. This new approach emphasizes modularity, testability, and enhanced performance

## Technical Details
New architecture
<img width="1967" height="720" alt="image" src="https://github.com/user-attachments/assets/e7f7df1e-6ad1-4224-bcfa-b598994c9761" />

## JIRA ID

AIPROFSYST-13

## Test Plan

<!-- Explain any relevant testing done to verify this PR. -->

## Test Result

<!-- Briefly summarize test outcomes. -->

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

sputhala-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi/processor.hpp:
```diff
@@ -0,0 +1,371 @@
+
+    void collect_gpu_metrics(smi_metrics& metrics) const
+    {
+        if(!m_supported_metrics.bits.vcn_activity &&
```
Not just supported we should also check if the settings is enabled by checking enabled.bits.vcn_activity etc. Current implementation collects these gpu metrics and filters during post-processing.

sputhala-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi/processor.hpp:
```diff
@@ -0,0 +1,371 @@
+        auto status = m_driver_api->get_power_info(m_processor_handle, &power_info);
+
+        m_supported_metrics.bits.current_socket_power =
+            power_info.current_socket_power != METRIC_VALUE_NOT_SUPPORTED &&
```
Is it good to have a helper function to check for MAX values :
```
template <typename T>
constexpr T metric_unsupported_value = std::numeric_limits<T>::max();

template <typename T>
constexpr bool is_metric_supported(T value) noexcept
{
    return value != metric_unsupported_value<T>;
}
```

sputhala-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi/amd_smi_sample.hpp:
```diff
@@ -0,0 +1,86 @@
+    {}
+
+    amd_smi::enabled_metric enabled_metric;
+    size_t                  device_id;
```
should this be also unint32_t to be consistent with other places and amd_smi.

sputhala-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi.hpp:
```diff
@@ -71,87 +54,15 @@ post_process();
 
-// Fork handling - cleanup AMD SMI state in child process
+uint32_t
+device_count();
```
This function is never used.

sputhala-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi/processor.hpp:
```diff
@@ -0,0 +1,371 @@
+
+    [[nodiscard]] smi_metrics get_smi_metrics() const
+    {
+        smi_metrics metrics{};
```
We initially discussed using amdsmi_get_gpu_metrics_info() to collect all metrics if most of them are enabled instead of calling individual APIs. Do we want to do that change here?

sputhala-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi/perfetto_policy.hpp:
```diff
@@ -0,0 +1,525 @@
+        }
+    }
+
+    static void post_process(size_t device_index, enabled_metric enabled_metrics,
```
Why are we bypassing the trace_cache for amd_smi sample tracing in perfetto?

 trace_cache::get_buffer_storage().store(trace_cache::amd_smi_sample{...});

sputhala-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi/common.hpp:
```diff
@@ -0,0 +1,201 @@
+    struct xcp_metrics
+    {
+        std::array<uint16_t, ROCPROFSYS_MAX_NUM_JPEG_ENGINES> jpeg_busy;
+        std::array<uint16_t, AMDSMI_MAX_NUM_XCP>              vcn_busy;
```
This should be `AMDSMI_MAX_NUM_VCN` ?

sputhala-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi/common.hpp:
```diff
@@ -0,0 +1,201 @@
+{
+    struct xcp_metrics
+    {
+        std::array<uint16_t, ROCPROFSYS_MAX_NUM_JPEG_ENGINES> jpeg_busy;
```
Do we want to remove support for `VCN_ACTIVITY` and `JPEG_ACTIVITY `on older ASICS?

adjordje-amd replied:
Yes, it is supported now

sputhala-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi/tests/tests_service.cpp:
```diff
@@ -0,0 +1,207 @@
+}  // namespace amd_smi
+}  // namespace rocprofsys
+
+#endif  // ROCPROFSYS_USE_ROCM > 0
```
The types.hpp file is blank. Possibly created by error - projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi/types.hpp

sputhala-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi/processor.hpp:
```diff
@@ -0,0 +1,371 @@
+        {
+            for(size_t xcp = 0; xcp < AMDSMI_MAX_NUM_XCP; ++xcp)
+            {
+                std::memcpy(metrics.xcp_stats[xcp].vcn_busy.data(),
```
Would `std:copy` be a better alternative here for logical element copying and explicit bounds ?

sputhala-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi/amd_smi_impl.hpp:
```diff
@@ -0,0 +1,181 @@
+                    case device_selection_mode::ALL: return processors;
+                    case device_selection_mode::NONE: break;
+                    case device_selection_mode::SPECIFIC:
+                        std::copy_if(processors.begin(), processors.end(),
```
Here we silently ignore invalid device indices specified by the user as parsed by `parse_numeric_range()`

sputhala-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi/service.hpp:
```diff
@@ -0,0 +1,174 @@
+                    m_driver_api->get_processor_type(processor_handle, &processor_type),
+                    "Failed to get processor type!");
+
+                processors.emplace_back(std::make_shared<processor_t>(
```
Shouldnt we check for `AMDSMI_PROCESSOR_TYPE_AMD_GPU` before adding?

sputhala-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi.cpp:
```diff
@@ -402,856 +69,72 @@ is_initialized()
-
     } catch(std::runtime_error& _e)
     {
         ROCPROFSYS_VERBOSE(0, "Exception thrown when initializing amd-smi: %s\n",
```
This should be `ROCPROFSYS_WARNING()`

sputhala-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/pmc/output_policies/cache_policy.hpp:
```diff
@@ -0,0 +1,294 @@
+        };
+
+        for(int xcp = 0; xcp < AMDSMI_MAX_NUM_XCP; ++xcp)
+        {
```
As specified somewhere else in the review comments we currently do not report `vcn_activity` and `jpeg_activity` values. I dont know if this is intentional, in which case we need confirmation from rocdecode/jpegdecode team if that should be fine because those values are reported only on older ASICs. But if those values are significant for the customers/users then we may have to take an informed decision.

sputhala-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/pmc/output_policies/cache_policy.hpp:
```diff
@@ -0,0 +1,294 @@
+              thread_id, "{}" });
+    }
+
+    static void initialize_smi_pmc_metadata(size_t gpu_id)
```
PMC metadata missing for VCN and JPEG?

sputhala-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi.hpp:
```diff
@@ -178,14 +89,22 @@ inline void
 {}
 
+inline uint32_t
+device_count()
```
Probably dead code?? I dont see this used anywhere.

sputhala-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi/common.hpp:
```diff
@@ -0,0 +1,201 @@
+{
+    std::stringstream ss;
+    ss << "[SMI enabled metrics] ";
+    ss << "Current socket power: " << metrics.bits.current_socket_power
```
`static_cast<bool>` ??

adjordje-amd replied:
yeah to output it as a bool (true, false)

sputhala-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi/common.hpp:
```diff
@@ -0,0 +1,201 @@
+#    ifdef AMDSMI_MAX_NUM_JPEG_ENG_V1
+#        define ROCPROFSYS_MAX_NUM_JPEG_ENGINES AMDSMI_MAX_NUM_JPEG_ENG_V1
+#    else
+#        define ROCPROFSYS_MAX_NUM_JPEG_ENGINES 40
```
Should this default to AMDSMI_MAX_NUM_JPEG like here: https://github.com/ROCm/rocm-systems/blob/7178747ebc310def89924ace32acc06a05a10737/projects/amdsmi/include/amd_smi/amdsmi.h#L146

sputhala-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi/amd_smi_impl.hpp:
```diff
@@ -0,0 +1,181 @@
+
+    void sample(const get_timestamp_t& get_timestamp)
+    {
+        auto _enabled_metrics = SettingsApi::get_enabled_metrics();
```
We are calling `get_enabled_metrics()` many times, should we instead use `m_enabled_metrics` ?

sputhala-amd commented on projects/rocprofiler-systems/source/lib/core/trace_cache/rocpd_processor.cpp:
```diff
@@ -337,192 +337,94 @@ rocpd_processor_t::handle([[maybe_unused]] const amd_smi_sample& _amd_smi)
+                           _amd_smi.device_id),
+                       enabled.jpeg_activity,
+                       [](const auto& xcp) -> const auto& { return xcp.jpeg_busy; });
+
```
XGMI metrics are collected but not processed here, its missing.

sputhala-amd commented on projects/rocprofiler-systems/source/lib/core/trace_cache/perfetto_processor.cpp:
```diff
@@ -925,275 +926,282 @@ perfetto_processor_t::handle([[maybe_unused]] const pmc_event_with_sample& _pmc)
 void
 perfetto_processor_t::handle([[maybe_unused]] const amd_smi_sample& _amd_smi)
 {
+    /*
```
Why are we not handling perfetto processing for GPU metrics here?
For testing we can still verify if trace_cache collects the data correctly, similar to how we should do for rocpd.
```
TEST(AmdSmiImpl, CollectsMetrics) {
    // Mock driver, verify metrics collected
    // Data goes to trace_cache (can verify)
}
```

sputhala-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi/common.hpp:
```diff
@@ -0,0 +1,201 @@
+
+constexpr auto METRIC_VALUE_NOT_SUPPORTED = 0xffff;
+
+union enabled_metric
```
The ordering of bit-fields within a `union` is implementation-defined. Different compilers or ABIs may lay out the bits differently, which makes this union non-portable and can result in unexpected behavior.

sputhala-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi/tests/tests_amd_smi_impl.cpp:
```diff
@@ -0,0 +1,316 @@
+    EXPECT_EQ(impl.get_processor_count(), 0u);
+}
+
+TEST_F(AmdSmiImplTest, DeviceFilterSpecificReturnsSelectedProcessors)
```
This is a good example of behavior testing - it verifies the observable outcome (which processors are returned) rather than checking internal mock state.

sputhala-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi/tests/tests_amd_smi_impl.cpp:
```diff
@@ -0,0 +1,316 @@
+    EXPECT_EQ(impl.get_processor_count(), 0u);
+}
+
+TEST_F(AmdSmiImplTest, SetupWithProcessorsInitializesPerfetto)
```
The tests are heavily focused on white-box testing (testing implementation) rather than black-box testing (testing behavior).

For example this verifies that setup() internally called init_storage() exactly once.

mradosav-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi/perfetto_policy.hpp:
```diff
@@ -0,0 +1,554 @@
+    static std::map<size_t, pcie_track_set> tracks;
+    return tracks;
+}
+
```
We can make this static variable inside perfetto_policy, as well as `get_perfetto_bundle`

mradosav-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi/processor.hpp:
```diff
@@ -0,0 +1,311 @@
+#if ROCPROFSYS_USE_ROCM > 0
+
+template <typename Driver>
+class processor
```
Processor looks like `gpu_metrics_source` or `gpu_device`

habajpai-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi.cpp:
```diff
@@ -1278,47 +165,27 @@ shutdown()
-device_count()
-{
-    return gpu::device_count();
+    printf("Post-processing amd-smi...\n");
```
Debug printf statement should be removed or replaced with ROCPROFSYS_VERBOSE before merge

habajpai-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi/perfetto_policy.hpp:
```diff
@@ -0,0 +1,554 @@
+
+#if ROCPROFSYS_USE_ROCM > 0
+
+namespace
```
Anonymous namespace in header - static state wont be shared if included from multiple .cpp files. Is single-TU usage intentional? Consider moving to .cpp or document the constraint.

habajpai-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/amd_smi/service.hpp:
```diff
@@ -0,0 +1,184 @@
+    service()
+    : m_driver_api(DriverFactory::create_driver())
+    {
+        check_status(m_driver_api->init(), "Failed to initialize AMD SMI driver!");
```
If init() succeeds but a later check_status() throws(eg. get_version() fails), the constructor fails and shutdown() is never called which could be potential AMD SMI driver state leak. 

I understand the double free issue was recently resolved(targetting 7.11 release), so this may not cause issues in practice. Just flagging for awareness - consider adding cleanup in catch block for reference.

adjordje-amd commented on projects/rocprofiler-systems/source/lib/core/trace_cache/tests/test_sample_type.cpp:
```diff
@@ -295,17 +295,17 @@ TEST_F(sample_type_test, pmc_event_with_sample_type_identifier)
 }
-
-TEST_F(sample_type_test, amd_smi_sample_serialize_deserialize)
+/*
```
TODO!

adjordje-amd commented on projects/rocprofiler-systems/source/lib/core/trace_cache/perfetto_processor.cpp:
```diff
@@ -115,18 +115,17 @@ using amd_smi_pcie_bandwidth_acc_track =
 using amd_smi_pcie_bandwidth_inst_track =
     perfetto_counter_track<category::amd_smi_pcie_bandwidth_inst>;
 
+/*
```
TODO!

adjordje-amd commented on projects/rocprofiler-systems/source/lib/core/trace_cache/perfetto_processor.cpp:
```diff
@@ -1010,67 +1011,78 @@ perfetto_processor_t::handle([[maybe_unused]] const pmc_event_with_sample& _pmc)
+        return fmt::format("GPU [{}] {} (S)", _device_id, metric);
+    };
+
+    // Create and emit basic GPU activity metrics
```
```suggestion
```

adjordje-amd commented on projects/rocprofiler-systems/source/lib/core/trace_cache/perfetto_processor.cpp:
```diff
@@ -1133,78 +1145,30 @@ perfetto_processor_t::handle([[maybe_unused]] const amd_smi_sample& _amd_smi)
-                                     gpu_metrics.vcn_activity, std::nullopt);
-    }
-    else
+    // Insert VCN activity metrics - iterate through all XCP stats
```
```suggestion
```

adjordje-amd commented on projects/rocprofiler-systems/source/lib/core/trace_cache/perfetto_processor.cpp:
```diff
@@ -1133,78 +1145,30 @@ perfetto_processor_t::handle([[maybe_unused]] const amd_smi_sample& _amd_smi)
 
-    // Insert JPEG activity metrics
-    if(capabilities.flags.jpeg_is_device_level_only)
+    // Insert JPEG activity metrics - iterate through all XCP stats
```
```suggestion
```

adjordje-amd commented on projects/rocprofiler-systems/source/lib/core/trace_cache/perfetto_processor.cpp:
```diff
@@ -1133,78 +1145,30 @@ perfetto_processor_t::handle([[maybe_unused]] const amd_smi_sample& _amd_smi)
         }
     }
 
     // Insert XGMI metrics
```
```suggestion
```

adjordje-amd commented on projects/rocprofiler-systems/source/lib/core/trace_cache/perfetto_processor.cpp:
```diff
@@ -1213,30 +1177,58 @@ perfetto_processor_t::handle([[maybe_unused]] const amd_smi_sample& _amd_smi)
 
-        insert_xgmi_vector_metrics(category::amd_smi_xgmi_read_data{}, is_xgmi_enabled,
-                                   gpu_metrics.xgmi_read_data_acc);
+        // XGMI read/write data per link
```
```suggestion
```

adjordje-amd commented on projects/rocprofiler-systems/source/lib/core/trace_cache/perfetto_processor.cpp:
```diff
@@ -1213,30 +1177,58 @@ perfetto_processor_t::handle([[maybe_unused]] const amd_smi_sample& _amd_smi)
+        }
     }
 
     // Insert PCIe metrics
```
```suggestion
```

adjordje-amd commented on projects/rocprofiler-systems/source/lib/core/trace_cache/rocpd_processor.cpp:
```diff
@@ -366,195 +366,140 @@ rocpd_processor_t::handle([[maybe_unused]] const amd_smi_sample& _amd_smi)
+        info::annotate_with_device_id<category::amd_smi_memory_usage>(_gpu_pmc.device_id),
+        enabled.bits.memory_usage, m.memory_usage / 1024.0);
+
+    // Per-XCP VCN/JPEG busy metrics (MI300)
```
```suggestion
```

adjordje-amd commented on projects/rocprofiler-systems/source/lib/core/trace_cache/rocpd_processor.cpp:
```diff
@@ -366,195 +366,140 @@ rocpd_processor_t::handle([[maybe_unused]] const amd_smi_sample& _amd_smi)
+    insert_scalar(
+        trait::name<category::amd_smi_memory_usage>::value,
+        info::annotate_with_device_id<category::amd_smi_memory_usage>(_gpu_pmc.device_id),
+        enabled.bits.memory_usage, m.memory_usage / 1024.0);
```
Magic number

adjordje-amd commented on projects/rocprofiler-systems/source/lib/core/trace_cache/rocpd_processor.cpp:
```diff
@@ -366,195 +366,140 @@ rocpd_processor_t::handle([[maybe_unused]] const amd_smi_sample& _amd_smi)
+                       enabled.bits.jpeg_busy,
+                       [](const auto& xcp) -> const auto& { return xcp.jpeg_busy; });
+
+    // Device-level VCN/JPEG activity metrics (Radeon)
```
```suggestion
```

adjordje-amd commented on projects/rocprofiler-systems/source/lib/core/trace_cache/rocpd_processor.cpp:
```diff
@@ -366,195 +366,140 @@ rocpd_processor_t::handle([[maybe_unused]] const amd_smi_sample& _amd_smi)
+            _gpu_pmc.device_id),
+        enabled.bits.jpeg_activity, m.jpeg_activity);
+
+    // PCIe metrics
```
```suggestion
```

adjordje-amd commented on projects/rocprofiler-systems/source/lib/core/trace_cache/rocpd_processor.cpp:
```diff
@@ -366,195 +366,140 @@ rocpd_processor_t::handle([[maybe_unused]] const amd_smi_sample& _amd_smi)
+                  info::annotate_with_device_id<category::amd_smi_pcie_bandwidth_inst>(
+                      _gpu_pmc.device_id),
+                  enabled.bits.pcie, m.pcie.bandwidth.inst);
+/*
```
TODO!

adjordje-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/pmc/collectors/common/tests/test_collector_slice.cpp:
```diff
@@ -0,0 +1,189 @@
+namespace testing
+{
+
+// Mock collector implementations for testing type erasure
```
```suggestion
```

adjordje-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/pmc/collectors/common/tests/test_collector_slice.cpp:
```diff
@@ -0,0 +1,189 @@
+    EXPECT_TRUE(gpu_collector->setup_called);
+    EXPECT_TRUE(nic_collector->setup_called);
+
+    // Call config on all collectors
```
```suggestion
```

adjordje-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/pmc/collectors/common/tests/test_collector_slice.cpp:
```diff
@@ -0,0 +1,189 @@
+    EXPECT_TRUE(gpu_collector->config_called);
+    EXPECT_TRUE(nic_collector->config_called);
+
+    // Call sample on all collectors
```
```suggestion
```

adjordje-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/pmc/collectors/common/tests/test_collector_slice.cpp:
```diff
@@ -0,0 +1,189 @@
+    EXPECT_TRUE(gpu_collector->sample_called);
+    EXPECT_EQ(nic_collector->sample_count, 1);
+
+    // Call sample again to verify independent state
```
```suggestion
```

adjordje-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/pmc/collectors/common/tests/test_collector_slice.cpp:
```diff
@@ -0,0 +1,189 @@
+
+TEST_F(collector_slice_test, collector_slice_is_non_owning)
+{
+    // Create a slice
```
```suggestion
```

adjordje-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/rocprofiler-sdk.cpp:
```diff
@@ -22,6 +22,7 @@
 
 #include "core/rocprofiler-sdk.hpp"
 #include "api.hpp"
+#include "binary/analysis.hpp"
```
Remove ?

adjordje-amd commented on projects/rocprofiler-systems/source/lib/core/trace_cache/cache_manager.hpp:
```diff
@@ -26,9 +26,8 @@
 #include "core/trace_cache/sample_type.hpp"
 #include "core/trace_cache/storage_parser.hpp"
-
+#include "library/pmc/collectors/gpu/sample.hpp"
```
Remove?

adjordje-amd committed:
refactor(pmc): rewrite AMD SMI collector using policy-based design

Replace the monolithic AMD SMI implementation with a modular,
policy-based architecture for better testability and maintainability.

Key changes:
- Introduce collector/device/driver/provider template hierarchy
- Use compile-time policies for metrics, caching, and output
- Add proper separation of concerns (sampling, caching, output)
- Support multiple output formats (Perfetto, RocPD)
- Add comprehensive unit tests with GMock

Technical details:
- Policy-based design eliminates virtual call overhead in hot paths
- Metrics are now configurable via bitset (enabled_metrics)
- Cache policies control data retention and flushing
- Output policies handle format-specific serialization

Bug fixes included:
- Fix vcn_busy array size (AMDSMI_MAX_NUM_VCN not XCP)
- Fix bit position mismatch for xgmi/pcie metrics (12,13 not 10,11)
- Fix iterator invalidation in collector sample loop
- Add noexcept to simple getters
- Remove unused includes

adjordje-amd committed:
fix(perfetto): remove shadowing lambda in XGMI metrics handler

Remove redundant make_track_name lambda inside the XGMI metrics block
that was shadowing the outer declaration at line 998, fixing the
compilation warning about declaration shadowing.

adjordje-amd committed:
fix(perfetto): remove shadowing lambda in PCIe metrics handler

The make_track_name lambda was already defined at function scope
(line 998) but was unnecessarily redefined within the PCIe metrics
block (line 1201), causing a -Werror=shadow compilation failure.

Removed the redundant inner lambda definition since the outer
lambda is already in scope and has identical behavior.

adjordje-amd committed:
fix(pmc): replace C++20 designated initializers with C++17 syntax

The code used C++20 designated initializer syntax (.field = value)
which caused compilation errors with -std=c++17 -Werror=pedantic.

Changed from:
  enabled_metrics foo = { .value = expr };

To C++17 compatible:
  enabled_metrics foo;
  foo.value = expr;

Affected files:
- cache_policy.hpp: store_sample() function
- perfetto_policy.hpp: post_process_device() function

mradosav-amd commented on projects/rocprofiler-systems/source/lib/core/trace_cache/rocpd_processor.cpp:
```diff
@@ -356,207 +356,131 @@ rocpd_processor_t::handle([[maybe_unused]] const pmc_event_with_sample& _pmc)
+                  info::format_track_name<category::amd_smi_pcie_bandwidth_inst>(),
+                  enabled.bits.pcie, m.pcie.bandwidth.inst);
+    /*
+        //TODO: XGMI metrics
```
Todo, please fix it .

mradosav-amd commented on projects/rocprofiler-systems/source/lib/core/trace_cache/sample_processor.hpp:
```diff
@@ -21,9 +21,14 @@
 // SOFTWARE.
 
 #pragma once
+#include "common/defines.h"
```
why do we need this include?

adjordje-amd replied:
not sure, I have to double-check

adjordje-amd replied:
removed

marantic-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/pmc/output_policies/perfetto_policy.hpp:
```diff
@@ -0,0 +1,654 @@
+
+#include <vector>
+
+#define JOIN(...) ::timemory::join::join(__VA_ARGS__)
```
Could we start using `fmt::format` and `fmt::join` in the newly generated code? Should be a lot cleaner.

adjordje-amd replied:
Sure

adjordje-amd committed:
fix(pmc): check enabled bits before collecting GPU metrics

The previous implementation collected all hardware-supported GPU metrics
and filtered them during post-processing. This wasted AMD SMI API calls
for metrics that users disabled via ROCPROFSYS_AMD_SMI_METRICS.

Now device::get_gpu_metrics() accepts an enabled_metrics parameter and
checks both hardware support AND user settings before collecting each
metric category (power, temperature, activity, memory, VCN/JPEG, XGMI,
PCIe).

This addresses PR review feedback requesting enabled.bits checks at
collection time rather than post-processing.

Changes:
- Modified device::get_gpu_metrics() to accept enabled_metrics parameter
- Updated all collect_*_metrics() methods with dual checks
- Updated collector to pass m_enabled_metrics to get_gpu_metrics()
- Updated unit tests to pass enabled metrics parameter
- Cleaned up redundant comments in types.hpp and perfetto_policy.hpp

adjordje-amd committed:
fix(pmc): replace debug printf with LOG_DEBUG and add static_cast

Replace debug printf statement with LOG_DEBUG macro for consistency
with the rest of the codebase. Also add missing static_cast<bool> for
current_socket_power and average_socket_power in to_string() function
to ensure consistent boolean output formatting.

Changes:
- Replace printf with LOG_DEBUG in perfetto_policy.hpp
- Remove std::once_flag wrapper (not needed for LOG_DEBUG)
- Add static_cast<bool> for power metrics in types.hpp to_string()

adjordje-amd committed:
fix(pmc): skip non-GPU processor types during device enumeration

AMD SMI can enumerate multiple processor types (GPUs, CPUs, etc).
The previous implementation attempted to create GPU device objects
for all processor types, which could lead to incorrect device
initialization or crashes with non-GPU processors.

Now explicitly check for AMDSMI_PROCESSOR_TYPE_AMD_GPU before
creating device objects, ensuring only GPU devices are added to
the collection list.

This addresses PR review feedback to verify processor type before
adding to the device list.

adjordje-amd committed:
fix(pmc): warn about invalid GPU device indices

When users specify device indices via ROCPROFSYS_SAMPLING_GPUS that
don't exist (e.g., "0,5,99" when only 3 GPUs are present), the invalid
indices were silently ignored with no feedback to the user.

Now emit a warning for each requested device index that exceeds the
actual number of enumerated GPU devices, informing the user of the
valid range.

This addresses PR review feedback to properly handle invalid device
indices from parse_numeric_range() instead of silently ignoring them.

adjordje-amd committed:
refactor(pmc): address PR review feedback

Enable XGMI metrics in RocPD processor by uncommenting and fixing the TODO.
Used insert_device_level_metrics for XGMI read/write arrays instead of
insert_scalar for proper handling of array metrics.

Replace JOIN macro with fmt::format in perfetto_policy for cleaner string
formatting. The JOIN macro was a wrapper around timemory::join::join which
is less readable than modern fmt::format syntax.

Remove unnecessary common/defines.h include from sample_processor.hpp since
ROCPROFSYS_USE_ROCM is provided as a compile definition by CMake.

Changes:
- Enable XGMI link width/speed and read/write data metrics
- Replace JOIN macro with fmt::format for track names
- Remove redundant include of common/defines.h

adjordje-amd committed:
refactor(pmc): introduce unified collector with traits-based design

Extract common collector logic into base::collector template that can be
parameterized with device-specific traits. This eliminates code duplication
when adding new device types (NIC, CPU) and improves testability.

Changes:
- Add collectors/base/collector.hpp with unified collection logic
- Add collectors/base/traits_check.hpp for compile-time trait validation
- Add collectors/gpu/gpu_traits.hpp with GPU-specific customization points
- Simplify gpu/collector.hpp to a type alias (~320 lines -> ~68 lines)

The traits pattern enables:
- Compile-time polymorphism without virtual function overhead
- Clear interface requirements via static_assert
- Easy extension for new device types (NIC, CPU)

adjordje-amd committed:
refactor(pmc): remove gpu_pmc_sample type alias

Remove the legacy `gpu_pmc_sample` type alias from trace_cache namespace
and update all usages to use the fully qualified `pmc::gpu::sample` type.

This cleanup eliminates namespace pollution and makes the code more
explicit about which sample type is being used. The enum value
`type_identifier_t::gpu_pmc_sample` remains unchanged for serialization
compatibility.

adjordje-amd committed:
refactor(pmc): centralize metric bit positions in metric_descriptors.hpp

Previously, metric bit positions were hardcoded in multiple files:
- types.hpp (union bit fields)
- settings.hpp (parsing mask computation)
- perfetto_policy.hpp (make_metric_value calls)

This made adding or reordering metrics error-prone, as changes had
to be synchronized across 3+ files manually.

Key changes:
- Add metric_descriptors.hpp as single source of truth for all metrics
- Replace union-based enabled_metrics with class providing type-safe
  accessors via metric_id enum
- Add metric_masks namespace with pre-computed masks for each metric
- Add user_metric_aliases array for environment variable parsing
- Update all consumers to use new accessor methods (.gfx_activity()
  instead of .bits.gfx_activity)

The new design maintains raw value access (.value(), .set_value())
while providing named accessors and compile-time validation.

adjordje-amd committed:
fix(pmc): address Phase 1 blocking issues from PR review

Fix seven critical issues identified in the PR review document that
could cause data loss, silent failures, ODR violations, or performance
problems in the PMC collector.

Issues fixed:
- Add missing vcn_activity/jpeg_activity to sample serialization
- Link GPU collector tests to unit test executable
- Throw exception on AMD SMI read failure instead of silent empty return
- Replace brittle string comparison with compile-time traits check
- Move anonymous namespace state into struct to prevent ODR violations
- Remove static cache in get_enabled_metrics() for fork reinit support
- Cache supported metrics during enumeration to optimize sample() path

adjordje-amd committed:
fix(pmc): ensure device registry persists until post-processing completes

The previous implementation cleared device entries during shutdown(),
but post_process() runs after shutdown() and needs device information
to serialize counter tracks. This caused empty Perfetto output because
device data was already gone.

Changes:
- Add device registry in perfetto_policy to store device index and
  supported metrics during init_storage()
- Simplify post_process_perfetto signature (no longer needs device entries)
- Move cleanup from shutdown() to end of post_process() to ensure data
  is available when needed

adjordje-amd committed:
refactor(pmc): rename vcn_busy/jpeg_busy to xcp_vcn_activity/xcp_jpeg_activity

Rename per-XCP VCN/JPEG metrics to clearly distinguish them from
device-level metrics. The new naming convention explicitly indicates
these are per-XCP partition metrics.

Key changes:
- Rename metric_id::vcn_busy → xcp_vcn_activity
- Rename metric_id::jpeg_busy → xcp_jpeg_activity
- Add separate Perfetto track entries for per-XCP metrics
- Add new Perfetto categories for VCN/JPEG busy
- Clean up metric_descriptors.hpp documentation
- Simplify format_supported_metrics using fmt::format

sputhala-amd commented on projects/rocprofiler-systems/source/lib/core/trace_cache/rocpd_processor.cpp:
```diff
@@ -356,207 +357,136 @@ rocpd_processor_t::handle([[maybe_unused]] const pmc_event_with_sample& _pmc)
+                                                 : m.average_socket_power);
+    insert_scalar(trait::name<category::amd_smi_memory_usage>::value,
+                  info::format_track_name<category::amd_smi_memory_usage>(),
+                  enabled.memory_usage(), m.memory_usage / 1024.0);
```
This is converted to KB not MB. 

<img width="266" height="49" alt="Image" src="https://github.com/user-attachments/assets/cdd95f81-cb6c-4418-a771-c269661399a4" />

adjordje-amd replied:
Done

sputhala-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/pmc/collectors/gpu/device.hpp:
```diff
@@ -0,0 +1,377 @@
+            return;
+        }
+
+        populate_if_supported(metrics.xgmi.link.width, gpu_metrics.xgmi_link_width);
```
For XGMI/PCIe we check if the value reported by the driver is valid at collection time. For all other metrics we dont do a check for validity for other metrics at collection. Is that intentional?

adjordje-amd replied:
Done

sputhala-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/pmc/gpu/types.hpp:
```diff
@@ -0,0 +1,149 @@
+#    ifdef AMDSMI_MAX_NUM_JPEG_ENG_V1
+#        define ROCPROFSYS_MAX_NUM_JPEG_ENGINES AMDSMI_MAX_NUM_JPEG_ENG_V1
+#    else
+#        define ROCPROFSYS_MAX_NUM_JPEG_ENGINES 40
```
`ROCPROFSYS_MAX_NUM_JPEG_ENGINES` is not 40 on all gfx, its 32 on most and 40 on the newer archs, in which case why are we defaulting to 40 in case `AMDSMI_MAX_NUM_JPEG_ENG_V1` is not defined

adjordje-amd replied:
Done

adjordje-amd committed:
refactor(pmc): move enumerate_devices() to traits

The base collector contained GPU-specific enumeration logic (socket/processor
iteration, processor type filtering with AMD SMI status codes). This violates
separation of concerns - the base class should only contain generic logic
while device-specific enumeration belongs in concrete implementations.

Key changes:
- Move enumerate_devices() implementation from base collector to gpu_traits
- Add device_entry struct to gpu_traits (was in base collector)
- Add has_enumerate_devices compile-time trait check
- Move version logging to AMD SMI provider constructor
- Remove obsolete trait functions: get_processor_handles(),
  filter_by_processor_type(), expected_processor_type()
- Remove has_version_info trait (no longer needed)

This makes the base collector truly generic and allows future collectors
(NIC, CPU) to implement their own enumeration patterns.

adjordje-amd committed:
fix(pmc): address PR review comments and standardize copyright headers

Address multiple PR review comments:

- Fix memory_usage unit conversion from bytes to MB (was incorrectly
  converting to KB by dividing by 1024 instead of units::megabyte)
- Fix JPEG engine count fallback to use AMDSMI_MAX_NUM_JPEG (32)
  instead of hardcoded 40 for older AMD SMI versions
- Make metric collection consistent by using populate_if_supported()
  for all scalar metrics (power, temperature, activity), matching
  the defensive approach already used for XGMI/PCIe metrics
- Replace std::stringstream with fmt::format for cleaner string
  formatting in to_string() and check_status() functions

Also standardize copyright headers across all changed files to use
the short SPDX format:
  // Copyright (c) Advanced Micro Devices, Inc.
  // SPDX-License-Identifier:  MIT

mradosav-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/pmc/collectors/base/traits_check.hpp:
```diff
@@ -0,0 +1,121 @@
+{};
+
+template <typename Traits>
+struct has_enumerate_devices<
```
Are you sure this is correct? This looks like it checks for `::device` and `::supported metrics`, not for `enumerate_devices()`

adjordje-amd replied:
there is an updated version of `has_enumerate_devices`, and also, the comment with the explanation is update.

mradosav-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/pmc/collectors/gpu/gpu_traits.hpp:
```diff
@@ -0,0 +1,323 @@
+
+        auto   driver         = provider->get_driver();
+        auto   socket_handles = provider->get_socket_handles();
+        size_t index          = 0;
```
```cpp
struct index_incrementer
{
    index_incrementer(size_t& index) : m_index(index) {}
    ~index_incrementer() { index++; }
private:
    size_t& m_index;
}


// ...

index_incrementer _incrementer(index);

```
maybe to add something like this. Since index is incremented on every iteration.

adjordje-amd replied:
Agree, it's error prone. Will be addressed.

mradosav-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/pmc/collectors/base/collector.hpp:
```diff
@@ -0,0 +1,239 @@
+                        Traits::template store_perfetto_sample<PerfettoApi>(
+                            _device_id, _metrics, _timestamp);
+                    }
+                    return false;  // Keep device
```
Maybe to have something like this:

```cpp
constexpr bool keep_device = false;
constexpr bool remove_device = true;
// ...

                    return keep_device;

                    // ...

                    return remove_device;
```

To keep code to be self documenting.

adjordje-amd replied:
I got your point, but the comment is good enough, I think. Also, this is the standard stl algorithm api, so it's clear enough

mradosav-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys/library/pmc/output_policies/cache_policy.hpp:
```diff
@@ -0,0 +1,363 @@
+ *
+ * @see perfetto_policy for direct Perfetto trace output
+ */
+struct cache_policy
```
Is this gpu related struct or have general purpose?
vcn, xcp etc. are initialized here, so I'm thinking if we want to add something new, is this a place?

adjordje-amd replied:
General purpose as we discussed offline

mradosav-amd commented on projects/rocprofiler-systems/source/lib/rocprof-sys-user/rocprofiler-systems/categories.h:
```diff
@@ -67,6 +48,8 @@ extern "C"
         ROCPROFSYS_CATEGORY_AMD_SMI_MEMORY_USAGE,
         ROCPROFSYS_CATEGORY_AMD_SMI_VCN_ACTIVITY,
         ROCPROFSYS_CATEGORY_AMD_SMI_JPEG_ACTIVITY,
+        ROCPROFSYS_CATEGORY_AMD_SMI_VCN_BUSY,
+        ROCPROFSYS_CATEGORY_AMD_SMI_JPEG_BUSY,
```
```suggestion
    ROCPROFSYS_CATEGORY_AMD_SMI_VCN_BUSY,
    ROCPROFSYS_CATEGORY_AMD_SMI_JPEG_BUSY,
```
Is this mistake? Since we have duplicates.

adjordje-amd replied:
It's on purpose as we discussed offline

mradosav-amd reviewed (approved):
Clear separation of responsibilities, easy to read, classes do only one thing. Expect those few comments, everything else looks good.

adjordje-amd committed:
fix(pmc): properly detect template enumerate_devices and simplify iteration

The has_enumerate_devices trait was checking for device_entry fields
instead of the actual enumerate_devices method. Since it's a template
function, detecting it requires SFINAE with dummy type instantiation.

Also refactored enumerate_devices loop to use a lambda, ensuring index
is incremented in exactly one place instead of scattered across
multiple continue paths.
